src/lib/operators/join_hash/hash_functions.hpp

```diff
@@ -0,0 +1,556 @@
+    output_offset += segment->size();
+  }
+
+  // create histograms per chunk
+  histograms.resize(chunk_offsets.size());
+
+  std::vector<std::shared_ptr<AbstractTask>> jobs;
+  jobs.reserve(in_table->chunk_count());
+
+  for (ChunkID chunk_id{0}; chunk_id < in_table->chunk_count(); ++chunk_id) {
+    jobs.emplace_back(std::make_shared<JobTask>([&, chunk_id]() {
+      // Get information from work queue
+      auto output_offset = chunk_offsets[chunk_id];
+      auto output_iterator = elements->begin() + output_offset;
+      auto segment = in_table->get_chunk(chunk_id)->get_segment(column_id);
+
+      // prepare histogram
+      auto histogram = std::vector<size_t>(num_partitions);
+
+      resolve_segment_type<T>(*segment, [&, chunk_id, keep_nulls](auto& typed_segment) {
+        auto reference_chunk_offset = ChunkOffset{0};
+        auto iterable = create_iterable_from_segment<T>(typed_segment);
+
+        iterable.for_each([&, chunk_id, keep_nulls](const auto& value) {
+          if (!value.is_null() || keep_nulls) {
+            const Hash hashed_value = std::hash<HashedType>{}(type_cast<HashedType>(value.value()));
+
+            /*
+            For ReferenceSegments we do not use the RowIDs from the referenced tables.
+            Instead, we use the index in the ReferenceSegment itself. This way we can later correctly dereference
+            values from different inputs (important for Multi Joins).
+            */
+            if constexpr (std::is_same_v<std::decay<decltype(typed_segment)>, ReferenceSegment>) {
+              *(output_iterator++) =
+                  PartitionedElement<T>{RowID{chunk_id, reference_chunk_offset}, hashed_value, value.value()};
+            } else {
+              *(output_iterator++) =
+                  PartitionedElement<T>{RowID{chunk_id, value.chunk_offset()}, hashed_value, value.value()};
+            }
+
+            const Hash radix = hashed_value & mask;
+            ++histogram[radix];
+          }
+          // reference_chunk_offset is only used for ReferenceSegments
+          if constexpr (std::is_same_v<std::decay<decltype(typed_segment)>, ReferenceSegment>) {
+            reference_chunk_offset++;
+          }
+        });
+      });
+
+      if constexpr (std::is_same_v<Partition<T>, uninitialized_vector<PartitionedElement<T>>>) {  // NOLINT
+        // Because the vector is uninitialized, we need to manually fill up all slots that we did not use
+        auto output_offset_end = chunk_id < chunk_offsets.size() - 1 ? chunk_offsets[chunk_id + 1] : elements->size();
+        while (output_iterator != elements->begin() + output_offset_end) {
+          *(output_iterator++) = PartitionedElement<T>{};
+        }
+      }
+
+      histograms[chunk_id] = std::move(histogram);
+    }));
+    jobs.back()->schedule();
+  }
+
+  CurrentScheduler::wait_for_tasks(jobs);
+
+  const size_t table_size_considered_large = 100'000;
+  if (in_table->row_count() >= table_size_considered_large) {
```

Comment:
This looks like added extra work (two loops) - was that removed from somewhere else or how do we make sure that we don't do more than before?

Reply:
See above.